### ISSUE TITLE
Custom platforms got deleted

### ISSUE THREAD

**Issue by macmladen** [omega8cc/boa#1075]
This is really an emergency to me so title may not be the best. I'll post as I get new information.

Briefly, the base:
- I made a platform in `static`
- I used symlink to reuse code from platform ** Drupal 7.43.2 P.007**
- I have two more similar platforms that are still there, working

**Comment by macmladen** [comment]
Two custom platforms were deleted (fortunately, I found them under `/data/disk/o1/undo`not sure if that is the version removed) and site in third is not working now (I will need to look deeper)

I am now working to restore, luckily they are not too important but the occasion freaked me out.

Last night I was finishing the site for the client so that he can check that in the morning. I was looking at the other site checking some similar things, both under same platform) and they were normally working at 23:50 when I went sleeping (CEST time zone).

In the morning two platforms were deleted (moved) luckily, database is in place.

**Comment by macmladen** [comment]
Ah, It seems I found out (though still strange): I symlinked to  Drupal 7.43.2 P.007 platform that I deleted.

I need to repair and will close this and then report back.

**Comment by macmladen** [comment]
They were symlinked to `/data/all/000/core/drupal-7.43.2` but that is still there although that code as platform is deleted.

**Comment by omega8cc** [comment]
You are not expected to symlink anything within the directory tree with name **static**. You should also use the limited shell access for code management, and this would not even allow you to create symlinks.

**Comment by omega8cc** [comment]
Please replace these symlinks with local copy of core, or it will happen again.

**Comment by macmladen** [comment]
You are right. I should have my local copy and not rely on system, however that system looked smart to me so I tried to use it.

The problem lied in **assumptions** and that always lead to problems.

Last week I was suspecting some of my decisions and working to change that, however I am in the middle of a project so I decided to postpone it until finished.

Today I will "patch" my platform, e.g. use new core 7.50 and migrate properly to new, non-symlinked.

I felt this experience will be useful if someone else, a.k.a "wise guy" like me tries to do the same.

Fortunately, BOA is well thought out system so it didn't destroy suspicious zombie.

However, that also raised question for me **not** having a backup (**_rush_** is my excuse.. seriously?!)

**Comment by memtkmcc** [comment]
Ouch, good to know you were able to revert this..

**Comment by macmladen** [comment]
OK, what Americans would say, **the takeaway here is:**

1. Never try to be **_smart_** but play along: build your platform independently so it is self-contained and lives completely in expected location and that is `~/static`
2. If the platform becomes **damaged**, Ægir will move it without notice to `~/undo` while platform will still be listed along with all sites like nothing happened.
3. **Damaged** in my case was missing `profiles` directory that triggered the action so anything similar may do the same
4. You will know that when you site suddenly show as **Under construction** while they are regularly listed in **Sites** list
5. **Verifying** platform or sites will fail fast with message that everything is missing because it is not there anymore
6. To **restore** do the following:
  1. moved back the platforms from /data/disk/o1/undo back to /data/disk/o1/static
  2. moved back sites and platforms *.alias.drushrc.phpback to /data/disk/o1/.drush
  3. verify platforms and sites to be sure, especially if you tried verify before which failed, sites will reappear only when verified

If anyone face similar problem, I'd be glad to help, however, this is the essence.

**Comment by macmladen** [comment]
@omega8cc This is very aggressive action against something that might be slight mistake and it will trigger at some dead time in the night (02:00 AM( leaving sites out of reach for at least 6 or 7 hours.

What do you think to issue a warning as soon as check finds out fault platform and send mail to admin (o1 or whoever)? 

The problem here is that working platform and working sites were removed without notice and even after that removal, everything was still in Ægir GUI which added to confusion.

The platform was working and sites were working, just their status will report that their profile was missing.

My suggestion:

1. If removing platform/sites, Ægir should send mail with explanation what happened, why it happened and offer explanation how to revert the action and fix the problem.
2. Also when **zombifing** some sites, it should do the same, send mail, inform what and why it happened, where things are and how to fix.

There are always:

1. Server administrator
2. Octopus administrator
3. Site administrator

and each should be notified on what just happened.

After all if that is part of mistake, no one should be punished severely for some faulty move, right?

**Comment by omega8cc** [comment]
While this would never happen if you would follow the docs, there is a room for improvements, for sure.

However, notifications are not a cure. It can easily generate enough noise to get ignored and filtered as spam. Also, the only contact to receive system level notifications is server admin, no one else.

We could also improve the docs to better explain good and bad practices, but it also has limited usefulness, because people tend to never read docs, until it is too late -- like when something is already broken.

It is also not reasonable to expect that we will try to figure out all bad practices and warn to not use them.

The only truly reliable a good method to avoid such problems is to make these autonomous procedures even more failsafe, by default.

This is what we need to figure out.

**Comment by macmladen** [comment]
To be clear and honest, I completely agree with you.

There is no way anyone could have an idea of all bad practices, but once action is triggered there is some reason (maybe not `profile is deleted`, but `platform damaged, moved to undo see docs/recover on how to recover`) but always there is explanation when action is triggered, that is my point.

And for such aggressive action, user should be notified (server admin) who will take appropriate action as that could really hurt and, obviously, could be undone and fixed.

I know it is not in recommended procedures, but fails are almost exclusively in such situations where procedure is not followed appropriately, maybe by misunderstanding, maybe by missing some step and maybe deliberately because someone though it _could be a good idea_.
